New features, bug fixes, and new test case framework for 4.16.0

welchr opened the pull request:
## RAREMETAL

New features: 

- Heterogeneity statistics, mean/se/min/max allele frequency tracking (#20)
- Run single variant tests w/o aggregation tests (#24)
- Increase p-value precision for single variant tests, and optional log-pvalue output (#23)
- New test case framework (Catch2) + test cases (cc6edc2c1cd07e845ec7116c4ac0e955b7dd88a3)
- Travis CI and status checks on master/develop branches (#29)
- Dockerfile for local development / testing linux builds (ff5e1dc7a09bd52a083dc0dd6049f00ab0e9fc37)

Bug fixes:

- Incorrect handling of dosage counts when determining if variant is monomorphic (#4)
- Running in a region with `--range chr:start-end` limits output to that region for both aggregation tests *and* single variant tests, previously only aggregation tests were limited (#26)
- Fix compile error on linux (bc976dd8369903ecd786a7c131a748c3944e8025)

## RAREMETALWORKER

Bug fixes: 

- Program would terminate normally even though input file was not entirely read when a multi-allelic variant was present (#7)

welchr committed:
Avoid truncation from double to integer

Fixes #4

welchr committed:
Add more logging statements, primarily for filtered variants

welchr committed:
Latest savvy version

welchr committed:
Raise exception if savvy fails while reading file

Otherwise, the program will finish normally, and only the log file
(not the console) will have a message regarding the failure (though it
is not marked as an error, so it is hard to see.)

Fixes #7

welchr committed:
Disable unsupported argument

welchr committed:
Missed imports; remove debug prints

welchr committed:
Implement support for heterogeneity statistics

Partially implements #20

welchr committed:
Allow skipping burden tests

Logic was incorrect in that if no burden tests were requested, it would
still ask for a covariance matrix anyway

Fixes #24

welchr committed:
Fix heterogeneity stats

Patch for #20

welchr committed:
Switch to Catch2 test framework

welchr committed:
Refactor Meta to remove global state

welchr committed:
Allow skipping output when testing

welchr committed:
Fix test case for heterogeneity chisq values

Test for #20

welchr committed:
Refactor parsing score/cov file listing

welchr committed:
Refactor single variant p-value calc; added -log10(p)

Issue #23

welchr committed:
Fix test cases; add test for -log10(p) on extreme small p-value

Issue #23

welchr committed:
Remove global state on GroupFromAnnotation

welchr committed:
Additional comments on het stats

welchr committed:
Write out -log10 single variant p-values

welchr committed:
Write het p-values in -log10 scale

welchr committed:
Refactor file output, log file creation

welchr committed:
Don't need to skipOutput on tests after refactor

welchr committed:
Refactor meta prep

welchr committed:
Supporting code and initial test case for file I/O

welchr committed:
Fix precision of p-value (even without --logPvalue) + test cases

Closes #23

welchr committed:
Fix single variant meta-analysis reader to use column header instead of fixed column positions

welchr committed:
Rename variable (it isn't maf, it's alt freq)

welchr committed:
Extra AF stats: mean, SE, min, max

Implements #20

welchr committed:
Update reader for new AF stats in meta-analysis results file

welchr committed:
Test case for new AF stats

Closes #20

welchr committed:
Additional comments

welchr committed:
Allow testing only within a given chrom:start-end

This limits both single variant and group test results to only include the range specified by the user

Implements #26

welchr committed:
Test case for --range

Closes #26

welchr committed:
Data files for range test

welchr committed:
Additional data files for tests

welchr committed:
Fix regression from new tabix usage

welchr committed:
Update libStatGen to fix linux compile error and pin to commit

welchr committed:
Update savvy version

welchr committed:
Dockerfile and .dockerignore

welchr committed:
Travis CI

Closes #29

welchr committed:
Fix pooled alt AF parsing

welchr committed:
Support for comparing two single variant result sets read in from a file for equality

welchr committed:
Fix logP getting switched on without command line arg given

welchr committed:
Check equality of metadata from single variant reader

welchr committed:
Update tutorial expected results after patches

welchr committed:
Add first tutorial dataset to test suite

welchr committed:
Compare equality of group test datasets

welchr committed:
Check group p-val, effect

welchr committed:
Check SKAT, burden, VT in tutorial dataset

welchr committed:
Cleanup logs, test output

welchr committed:
Fix another linux compile error

Apparently clang will resolve isnan without std::, but gcc won't...

welchr committed:
Move file exists

welchr committed:
Modify all readers to check for file existence and successful record read

welchr committed:
Add other tutorial test case for rvtest input

welchr committed:
Fix regression in reading rvtest formatted files

welchr committed:
Load heterogeneity p-value in SV reader

welchr committed:
Patch heterogeneity code to obey --range

welchr committed:
Test case for heterogeneity and --range option

welchr committed:
Allow skipping PDF generation

welchr committed:
Fix PDF heisenbug

welchr committed:
Tests for burden, SKAT, VT under the alternative

welchr committed:
Test with a monomorphic variant

welchr commented:
@abought - did you want to look at anything in here before I merge into develop?

abought commented:
Hi Ryan! Thanks for all the (tremendous) work on this.

20,000 lines is quite a lot to take in all at once for feedback! 

Two initial speedbumps:
1. I'm having some trouble compiling on Mac OS 10.14: it seems to require a savvy upgrade but I get errors about `xz not found`. Do we need to document any dependency changes in the README?

2. At this instant in time, there doesn't seem to be a single consolidated definition of what constitutes "passing release criteria". Can we update the repo and docs?
  * Based on the current travis.yml, the old integration tests still exist in the repo, but aren't being run. This creates a grey area where they might be relevant- or not. Can we either verify they are passing (via CI), *or* remove the tests as a dead end? From the old README: `To run unit tests, ensure that you have run cmake with -DBUILD_TESTS=1 and built at least once, then run make or ctest --verbose.`  (** I think that should say "make test"?)

  * Based on travis.yml, there are some new commands required to run unit tests, and these commands are different from what is documented in the README.

jonathonl commented:
Apple decided (for whatever reason) to prevent /usr/include directories on recent versions of MacOS, which is breaking some autotools projects. The xz dependency actually is not required, and you should be able to use `cget ignore xz` to get around this.

welchr commented:
I've noticed a few times too that when cget runs and tries to download certain tarballs (like xz) it will randomly fail, and I just have to re-run it again. When you run `cget install -f requirements.txt` is it failing to grab xz, or is it failing to install entirely?

welchr committed:
Travis build for OSX

jonathonl commented on .travis.yml:
```diff
@@ -1,20 +1,41 @@
+      brew install cmake zlib xz python3
+      brew cask install gfortran
+      sudo python3 -m pip install cget
+      sudo installer -pkg /Library/Developer/CommandLineTools/Packages/macOS_SDK_headers_for_macOS_10.14.pkg -target /
```
This package is not available on MacOS Catalina. A better approach is to set CPATH to the SDK headers location before running cget:
`export CPATH=$(xcrun --show-sdk-path)/usr/include`

welchr replied:
I gave that a shot for 10.5, but I get a huge number of compiler warnings. The job ends up being killed because the log exceeds the maximum allowed size. I think you can see the log [on the travis site](https://travis-ci.com/github/statgen/raremetal/builds/168453071), it's for the Apple build without a version number (not sure why it's blank and not 10.5.) This might be a clang issue on Catalina but I have no way to test it unfortunately. 

When I tried inserting a quick debug `clang++ -E -x c++ - -v < /dev/null` to see the search paths, that path `$(xcrun --show-sdk-path)/usr/include` was already there (at least on Travis' 10.5 VM.)

Do you happen to have an example of using this in Travis, or a Catalina box available to test with? I'm still running on Mojave and don't plan to upgrade anytime soon.

jonathonl replied:
You can silence the excessive compiler warnings with `export CFLAGS="-Wno-nullability-completeness -Wno-expansion-to-defined -Wno-undef"`. I'm assuming no version means the latest version. I've never targeted anything other than linux on Travis, but I have Catalina on one of my laptops and tested the warning suppression above. I didn't know there was a max log size, but I suppose that makes sense. 

Don't upgrade upgrade if you don't have to. It's not worth the headache ;).

welchr committed:
Support both MacOS 10.14 and 10.15 in travis

welchr committed:
Remove old raremetal tests

raremetalworker tests will be ported over later if we choose to continue
to support it.

welchr committed:
Fix raremetalworker binary path in cmake tests

welchr committed:
Update README

welchr committed:
Run RMW tests on OSX

welchr commented:
Updates so far: 

* I think we've got the compiling issues sorted out, and travis is updated to test for Mac OS 10.14 and 10.15 (thank you Jonathon for all the help here. 👍) I added some helpful info to the README about this as well.

* The older raremetal test cases are removed, and I updated the README with the relevant test commands for running the new test suite.

One lingering problem, unfortunately:

* I created a branch where I tried adding in the old raremetalworker test cases to travis. Unfortunately, one of the older raremetalworker test cases [only passes on MacOS](https://travis-ci.com/github/statgen/raremetal/builds/168643560), but not on Linux. There is a slight numerical precision difference on the two platforms and that causes enough of a diff that it fails (I confirmed this with a Linux docker image outside of Travis.)

  I think I'm inclined to let this go, since there's a fair chance that raremetalworker will be deprecated in favor of rvtest in the future, and this PR doesn't change the status quo at least (the test would never have worked in Linux, and no code changes here touch raremetalworker code.)

  To fix it would require the same work needed in raremetal - refactor everything to remove global state(s), add in supporting code for loading and comparing results, and then add in the test cases as well. That's quite a bit of work. I'm not sure it would be time well spent at this point, either.

  What do you guys think?

abought commented:
I can confirm that the precision difference is the cause of the testcase failure: this was one of the big drawbacks to the old diff-output-files approach. Very glad to see real tests (and Travis) added; thanks both!

If we actively plan to remove RAREMETALWORKER, then getting rid of those tests is a plausible option.  I don't have as much experience with ctest, but I assume that the tests could conditionally only be run on a specific platform?

With our unpredictable schedules, we don't always get around to removing code as planned, and then years later a student assumes it's safe to change. If we could keep the tests (where relevant) without too much effort, this might be a middle ground where we avoid the drawbacks, but keep the safety net.

Either way: when the code really is removed, go wild and remove the dead tests too.

welchr commented:
Thanks Andy that made good sense to me. I made the old raremetalworker tests only run under OSX in Travis, so that should prevent future build failures. They're still in the repo too in case we end up potentially working on raremetalworker in the future.

abought commented:
As reported in an email from Sarah Graham:

> I did catch a bug in this version (fix/welchr), when the condition option is used, the header gets the last column twice:
"#CHROM  POS     REF     ALT     N       POOLED_ALT_AF   DIRECTION_BY_STUDY      EFFECT_SIZE     EFFECT_SIZE_SD  H2      PVALUE  COND_EFFSIZE    COND_EFFSIZE_SD COND_H2 COND_PVALUE     COND_PVALUE"

welchr committed:
Fix missed columns in SKAT result header when --condition used

welchr committed:
Fix COND_PVALUE repeated twice in header

welchr committed:
Expose headers in group test and single variant readers

welchr committed:
Read in conditional effect size, p-value for single variants

welchr committed:
Add test case for --condition

welchr committed:
Fix incorrect dimension for temporary X^T * X matrix

This can end up causing eigen to ask for a very large amount of unnecessary memory if sample size is large

welchr committed:
Disable --useCovariates as projection matrix is never used

welchr committed:
Fix issue with zlib

welchr committed:
Protect against seg fault caused by ifeof() bug
